core/che-core-api-core/src/test/java/org/eclipse/che/api/core/util/lineconsumer/ConcurrentFileLineConsumerTest.java

```diff
@@ -64,18 +89,82 @@ public void shouldBeAbleToWriteIntoFile() throws Exception {
     @Test
     public void shouldNotWriteIntoFileAfterConsumerClosing() throws Exception {
         // given
-        injectWriterMock(concurrentFileLineConsumer, writer);
-        doNothing().when(writer).write(anyString());
-        doNothing().when(writer).flush();
+        final String message = "Test line";
 
         // when
         concurrentFileLineConsumer.close();
-        concurrentFileLineConsumer.writeLine("Test line");
+        concurrentFileLineConsumer.writeLine(message);
 
         // then
         verify(writer, never()).write(anyString());
     }
 
+    @Test
+    public void shouldBeAbleToWriteIntoFileSimultaneously() throws Exception {
+        // given
+        final String message1 = "Message 1";
+        final String message2 = "Message 2";
+
+        WaitingAnswer<Void> waitingAnswer = new WaitingAnswer<>();
+        doAnswer(waitingAnswer).when(writer).write(eq(message1));
+
+        executor.execute(() -> writeIntoConsumer(message1));
+        waitingAnswer.waitAnswerCall(1, TimeUnit.SECONDS);
+
+        // when
+        writeIntoConsumer(message2);
+        waitingAnswer.completeAnswer();
+
+        // then
+        awaitFinalization();
+
+        verify(writer).write(eq(message1));
+        verify(writer).write(eq(message2));
+    }
+
+    @Test
+    public void closeOperationShouldWaitUntilAllCurrentWriteOperationsWillBeFinished() throws Exception {
+        // given
+        final String message1 = "Message 1";
+        final String message2 = "Message 2";
+
+        WaitingAnswer<Void> waitingAnswer1 = waitOnWrite(message1);
+        WaitingAnswer<Void> waitingAnswer2 = waitOnWrite(message2);
+
+        // when
+        executor.execute(this::closeConsumer);
+
+        waitingAnswer1.completeAnswer();
+        waitingAnswer2.completeAnswer();
+
+        // then
+        awaitFinalization();
+
+        assertFalse(concurrentFileLineConsumer.isOpen());
+
+        verify(writer).write(eq(message1));
+        verify(writer).write(eq(message2));
+    }
+
+    @Test
+    public void shouldIgnoreWriteOperationsWhenLockForCloseIsLocked() throws Exception {
+        // given
+        WaitingAnswer<Void> waitingAnswer = new WaitingAnswer<>();
+        doAnswer(waitingAnswer).when(writer).close();
+
+        executor.execute(this::closeConsumer);
+        waitingAnswer.waitAnswerCall(1, TimeUnit.SECONDS);
+
+        // when
+        writeIntoConsumer("Test line");
```

Comment:
I think, that it is OK. IMO, good test shouldn't know about an implementation, but be able to verify correct work of functionality. Here we shouldn't care about implementation behaviour in the test case (do nothing, log warning/error or throw an exception). All what we need to know that method which prints message in subconsumer wasn't invoked.